### PR TITLE
mruby-regexp: let an empty group take a quantifier

### PR DESCRIPTION
**Pull request by takumin**
An empty non-capturing group takes no quantifier: `(?:)` followed by `*`,
`+`, `?` or an interval raises `target of repeat operator is not specified`,
where CRuby compiles the group as an empty match and the quantifier as one of
it. The scoped-option spelling `(?i:)` and a plain `()` demoted by a named
group are refused the same way, while `()`, `(?:|)` and `(?=)` take the
quantifier on both sides:

```ruby
/(?:)*/ =~ ""          # CRuby: 0,   mruby: RegexpError, target of repeat operator is not specified
/(?:)+/ =~ ""          # CRuby: 0,   mruby: RegexpError
/(?:){2}/ =~ ""        # CRuby: 0,   mruby: RegexpError
/(?:(?:))*/ =~ ""      # CRuby: 0,   mruby: RegexpError
/a(?:)*b/ =~ "ab"      # CRuby: 0,   mruby: RegexpError
/(?i:)*/ =~ ""         # CRuby: 0,   mruby: RegexpError
/(?<n>a)()*/ =~ "a"    # CRuby: 0,   mruby: RegexpError
/()*/ =~ ""            # 0 in both
/(?:|)*/ =~ ""         # 0 in both
/(?=)*/ =~ ""          # 0 in both
```

`compile_atom()` emits nothing for the empty group, so `compile_quantified()`
sees no atom and returns, and the quantifier is left to `compile_seq()`, whose
guard for a quantifier with no atom refuses it. `()` emits its two `RE_SAVE`s
and `(?:|)` its `RE_SPLIT`, so those have an atom.

## The fix

`compile_atom()` returns whether it read an atom, which an empty group is and
an option toggle `(?i)` is not, and `compile_quantified()` reads the
quantifiers of an atom that emitted nothing the way it already reads those of
`a{0}`, with `skip_quantifiers()`: a repeat of what matches empty matches
empty, so they emit nothing. `(?i)*` and a stray metacharacter still reach the
guard and are refused, as CRuby refuses them:

```ruby
/(?i)*/     # RegexpError in both
/(?#c)*/    # RegexpError in both
/(?:){a}/   # matches "{a}" in both: a `{` that spells no quantifier is a literal
```

## Size

`.text` of `bin/mruby`, `build_config/ci/gcc-clang.rb`, each side from a clean
build directory. `re_compile.o` is the only object that changes: by 41 bytes
at `-O0`, and at `-O3` `skip_quantifiers()` has two callers now and gcc stops
inlining it into `compile_seq()` (298 bytes in `bintest`), the rest being the
layout of `compile_seq()`, into which the parser is inlined.

| build | master | this PR | delta |
|---|---:|---:|---:|
| `bintest` | 1,281,398 | 1,281,846 | +448 |
| `ascii-ctype` | 1,269,286 | 1,269,766 | +480 |
| `byte-string` | 1,251,142 | 1,251,958 | +816 |
| `cxx_abi` | 1,306,873 | 1,307,097 | +224 |
| `full-debug` (`-O0`) | 1,881,126 | 1,881,174 | +48 |

## Verification

The tests go in `regexp_syntax.rb` beside the non-capturing group test: the
quantifiers above on `(?:)`, `(?i:)` and a demoted `()`, the lazy, possessive
and stacked spellings, `(?:) *` under `/x`, a `{` that spells no quantifier,
and `(?i)*` still refused. On master the block fails at its first assertion,
the literal being refused when it is compiled.

**Differential against CRuby 4.0.6**, the harness of #7269 and #7273 with
master and this PR against the same cases: 10,000 random patterns (seed 2,
the default features, over `a` and `b`), each matched against one subject and
compared as `MatchData#to_a`. Master refuses 1,657 of them, every one for
this reason, and CRuby cannot finish 1 under the memory cap. This PR refuses
none. Of the 1,657, 1,631 answer as CRuby does; 21 read a capture written
inside a lookaround, which is #7273 (with its fix stacked on this PR all 21
answer as CRuby does); and 5 are standing differences of the engine that the
empty group has no part in: each of the 5 answers the same, on master as
here, with its empty group and quantifier removed. The 8,342 cases both sides
compile answer the same on master and this PR, 100 of them differing from
CRuby on both.

**`rake test`**, `build_config/ci/gcc-clang.rb`, no compiler warning:

| build | total | KO | crash |
|---|---:|---:|---:|
| `full-debug` | 2,366 | 0 | 0 |
| `bintest` | 2,366 | 0 | 0 |
| `bintest` (bintest suite) | 123 | 0 | 0 |
| `cxx_abi` | 2,366 | 0 | 0 |
| `byte-string` | 2,295 | 0 | 0 |
| `ascii-ctype` | 2,362 | 0 | 0 |

The default configuration: 2,141 total, 0 KO, 0 crash, plus its 112 bintests.

### Environment

<details>

| | |
|---|---|
| OS | Ubuntu 24.04, Linux 7.0.0 x86_64, AMD Ryzen 9 5950X |
| gcc | 13.3.0 (Ubuntu 13.3.0-6ubuntu2~24.04.1) |
| binutils | 2.47 |
| CRuby | 4.0.6, for the differential |

Compile lines for `mrbgems/mruby-regexp/src/re_compile.c` in the builds
quoted above, paths shortened:

```
# ci/gcc-clang full-debug
gcc -MMD -c -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -g3 -O0 -DMRB_GC_STRESS -DMRB_USE_DEBUG_HOOK -DMRBGEM_MRUBY_REGEXP_VERSION=0.0.0 -DMRB_DEBUG -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER -I"include" -I"mrbgems/mruby-regexp/include" -I"build/full-debug/include" -o "build/full-debug/mrbgems/mruby-regexp/src/re_compile.o" "mrbgems/mruby-regexp/src/re_compile.c"

# ci/gcc-clang bintest
gcc -MMD -c -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_GC_FIXED_ARENA -DMRBGEM_MRUBY_REGEXP_VERSION=0.0.0 -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER -DMRB_USE_DEBUG_HOOK -I"include" -I"mrbgems/mruby-regexp/include" -I"build/bintest/include" -o "build/bintest/mrbgems/mruby-regexp/src/re_compile.o" "mrbgems/mruby-regexp/src/re_compile.c"

# ci/gcc-clang cxx_abi
gcc -MMD -c -g -O3 -Wall -Wundef -Wwrite-strings -x c++ -std=gnu++03 -DMRB_GC_FIXED_ARENA -DMRBGEM_MRUBY_REGEXP_VERSION=0.0.0 -DMRB_USE_CXX_EXCEPTION -DMRB_USE_CXX_ABI -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER -I"include" -I"mrbgems/mruby-regexp/include" -I"build/cxx_abi/include" -o "build/cxx_abi/mrbgems/mruby-regexp/src/re_compile.o" "mrbgems/mruby-regexp/src/re_compile.c"

# ci/gcc-clang byte-string
gcc -MMD -c -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRBGEM_MRUBY_REGEXP_VERSION=0.0.0 -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER -I"include" -I"mrbgems/mruby-regexp/include" -I"build/byte-string/include" -o "build/byte-string/mrbgems/mruby-regexp/src/re_compile.o" "mrbgems/mruby-regexp/src/re_compile.c"

# ci/gcc-clang ascii-ctype
gcc -MMD -c -std=gnu99 -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_USE_ASCII_CTYPE -DMRBGEM_MRUBY_REGEXP_VERSION=0.0.0 -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER -I"include" -I"mrbgems/mruby-regexp/include" -I"build/ascii-ctype/include" -o "build/ascii-ctype/mrbgems/mruby-regexp/src/re_compile.o" "mrbgems/mruby-regexp/src/re_compile.c"
```

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed regular expression handling for quantifiers applied to empty groups.
  * Improved support for greedy, lazy, possessive, nested, scoped-option, and bounded repetitions.
  * Preserved existing errors for invalid quantifier targets and option-toggle patterns.
  * Corrected handling of literal braces that are not valid quantifiers.

* **Tests**
  * Added regression coverage for empty-group quantification and related syntax edge cases.

### CODE REVIEW

File: mrbgems/mruby-regexp/src/re_compile.c [hunk 1]
```diff
@@ -1815,10 +1820,19 @@ compile_quantified(re_compiler *c)
...
   c->atom_start = saved_atom_start;
-  if (c->code_len == begin) return;  /* no atom emitted */
+  if (c->code_len == begin) {
+    /* Nothing emitted. An empty group `(?:)` is an atom all the same, and
+       one that matches empty matches empty however it is repeated, so its
+       quantifiers are read and emit nothing, as those of `a{0}` are; CRuby
+       compiles `(?:)*` the same way. What is not an atom, an option toggle
+       `(?i)` or a stray metacharacter, leaves a quantifier after it for
+       compile_seq() to refuse, as CRuby refuses `(?i)*`. */
+    if (atom) skip_quantifiers(c);
```
[the remark this answers is not in the document]
Comment (takumin) [reply mrbgems/mruby-regexp/src/re_compile.c:1833]: Not this PR: `parse_quantifier()` accepts a reversed interval everywhere already. On master `Regexp.new("a{2,1}")`, `Regexp.new("(a){2,1}")` and `Regexp.new("a{0}{2,1}")` all compile (as `{2}`), where CRuby raises `upper is smaller than lower in repeat range`; this PR only lets `(?:){2,1}` reach the same reader `a{2,1}` goes through, and changes nothing about what that reader accepts. The bounds check is a standing difference of the interval parser and belongs with the other interval-parser findings, not with the empty group.
